aten/src/ATen/cpu/vec/vec512/vec512.h

```diff
@@ -0,0 +1,224 @@
+ }
+
+template <typename T>
+std::ostream& operator<<(std::ostream& stream, const Vectorized<T>& vec) {
+  T buf[Vectorized<T>::size()];
+  vec.store(buf);
+  stream << "vec[";
+  for (int i = 0; i != Vectorized<T>::size(); i++) {
+    if (i != 0) {
+      stream << ", ";
+    }
+    stream << buf[i];
+  }
+  stream << "]";
+  return stream;
+}
+
+
+#if defined(CPU_CAPABILITY_AVX512) && !defined(_MSC_VER)
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ CAST (AVX) ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template<>
+inline Vectorized<float> cast<float, double>(const Vectorized<double>& src) {
+  return _mm512_castpd_ps(src);
+}
+
+template<>
+inline Vectorized<double> cast<double, float>(const Vectorized<float>& src) {
+  return _mm512_castps_pd(src);
+}
+
+#if defined(CPU_CAPABILITY_AVX512)
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ CAST (AVX2) ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#define DEFINE_FLOAT_INT_CAST(int_t, float_t, float_ch)                   \
+template<>                                                                \
+inline  Vectorized<int_t> cast<int_t, float_t>(const Vectorized<float_t>& src) {  \
+  return _mm512_castp ## float_ch ## _si512(src);                         \
+}                                                                         \
+template<>                                                                \
+inline Vectorized<float_t> cast<float_t, int_t>(const Vectorized<int_t>& src) {   \
+  return _mm512_castsi512_p ## float_ch (src);                            \
+}
+
+DEFINE_FLOAT_INT_CAST(int64_t, double, d)
+DEFINE_FLOAT_INT_CAST(int32_t, double, d)
+DEFINE_FLOAT_INT_CAST(int16_t, double, d)
+DEFINE_FLOAT_INT_CAST(int64_t, float, s)
+DEFINE_FLOAT_INT_CAST(int32_t, float, s)
+DEFINE_FLOAT_INT_CAST(int16_t, float, s)
+
+#undef DEFINE_FLOAT_INT_CAST
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ GATHER ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template<int64_t scale = 1>
+std::enable_if_t<scale == 1 || scale == 2 || scale == 4 || scale == 8, Vectorized<double>>
+inline gather(const double* base_addr, const Vectorized<int64_t>& vindex) {
+  return _mm512_i64gather_pd(vindex, base_addr, scale);
+}
+
+template<int64_t scale = 1>
+std::enable_if_t<scale == 1 || scale == 2 || scale == 4 || scale == 8, Vectorized<float>>
+inline gather(const float* base_addr, const Vectorized<int32_t>& vindex) {
+  return _mm512_i32gather_ps(vindex, base_addr, scale);
+}
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ MASK GATHER ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template<int64_t scale = 1>
+std::enable_if_t<scale == 1 || scale == 2 || scale == 4 || scale == 8, Vectorized<double>>
+inline mask_gather(const Vectorized<double>& src, const double* base_addr,
+                   const Vectorized<int64_t>& vindex, const Vectorized<double>& mask) {
+  auto all_ones = _mm512_castsi512_pd(_mm512_set1_epi64(0xFFFFFFFFFFFFFFFF));
+  auto mask_ = _mm512_cmp_pd_mask(all_ones, mask.values, _CMP_EQ_OQ);
+  return _mm512_mask_i64gather_pd(src, mask_, vindex, base_addr, scale);
+}

```

Comment:
Sorry @ezyang, but I'm a bit confused. What do comments such as this one signify?
Are you trying to evaluate the possibility of templatizing all AVX2/AVX512 files by using diff?
Thanks!

Reply:
The AVX512 intrinsic for `mask_i64gather` requires its mask to be a 8-bit bitmask, while AVX2 requires it to be `__m256d`.
Hence this snippet of code is faster for AVX2.
However, one way to mitigate this situation might be to use conditionals to ensure to that AVX2 & AVX512 code can have different arguments in calls to `mask_gather`. I'll see if that's a possibility. I'll get back to you tomorrow.

Reply:
For AVX512, I can make these masks bitmasks (eg. __mmask16, etc) instead of vector-masks.

https://github.com/pytorch/pytorch/blob/c830db02653a5ab555d966a3252426cfe8b6737a/aten/src/ATen/native/cpu/GridSamplerKernel.cpp#L515-L526

AVX512_256 also stands to benefit from such an optimization, but currently, I'm not using different intrinsics for it, so it won't.

Reply:
If I'd optimize it, memory-consumption would decrease a bit as bitmasks would used be instead of vector-width masks (If integer type is int8_t, then the LSB in each group of 8 bits in a 512-bit Vector is used) .
However, computing the bitmask would result in more computations.

eg. Let's consider Line 524 above
`auto ne_mask = cast<scalar_t>(e_mask & n_mask);`

In this case, the bitmask would've to be generated with something like -
```
__mmask64 inline and_mask(const Vectorized<int8_t>& a, const Vectorized<int8_t>& b) {
  auto c = a & b;
  auto d = Vectorized<int8_t>(1);
  // we're concerned with LSBs
  auto e = c & d;
  return _mm512_cmp_epi8_mask(d, e, _MM_CMPINT_EQ);
}
```
Also, using the bitmask would require an extra right-shift.

So, for now, I think I shouldn't change `mask_gather`. Thanks!